### PR TITLE
[WIP] Validate interdependencies in InterDependencies object

### PR DESCRIPTION
**Pull request by WilliamHPNielsen**
This is a yak-shave PR 🙂 

In order to validate array parameter interdependencies when adding results (as pointed out by @jenshnielsen in #1430), I want to refactor all the inter-dependency validation out of the datasaver and into a dedicated object.

So far, this PR does it for `register_parameter`. Next step: `add_result`. After that refactor, I believe that it will be much easier to add the length check to `add_result` without screwing up something.

Changes proposed in this pull request:
- Extend the `InterDependencies` object to do validation. The validation is done in the `__init__` to avoid confusion by mutation
- Refactor all validation from the `measurements.py` module into use of that object

@QCoDeS/core

### CODE REVIEW

**Comment by jenshnielsen** [comment]
The failing test uncovers a bug in the upgrade logic. I am looking into that at the moment

File: qcodes/dataset/dependencies.py [hunk 1]
```diff
@@ -39,6 +64,141 @@ def serialize(self) -> Dict[str, Any]:
...
+
+        needed: Set[str] = set()
+        present = set()
+
+        for param in params:
+            if param.name in needed:
+                needed.remove(param.name)
+            present.add(param.name)
+            param_deps = [sp for sp in param.depends_on.split(', ')
+                          if sp != '']
```
[the remark this answers is not in the document]
Comment (astafan8) [reply qcodes/dataset/dependencies.py]: now we have, it is `depends_on_` :)

[the remark this answers is not in the document]
Comment (jenshnielsen) [reply qcodes/dataset/dependencies.py]: Thanks, I remember seeing it somewhere :)

**Comment by jenshnielsen** [comment]
Hopefully #1453 has resolved the issues seen in the tests here

**Comment by WilliamHPNielsen** [comment]
Closing this in favour of #1477

**Pull request closed without merging**